akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -416,17 +417,22 @@ private[http] object HttpServerBluePrint {
             log.warning(
               """Sending 2xx response before end of request was received...
                 |Note that the connection will be closed after this response. Also, many clients will not read early responses!
-                |Consider waiting for the request end before dispatching this response!""".stripMargin)
+                |Consider only issuing this response after the request data has been completely read!""".stripMargin)
           val close = requestStart.closeRequested ||
-            requestStart.expect100Continue && oneHundredContinueResponsePending ||
-            isClosed(requestParsingIn) && openRequests.isEmpty ||
+            (requestStart.expect100Continue && oneHundredContinueResponsePending) ||
+            (isClosed(requestParsingIn) && openRequests.isEmpty) ||
             isEarlyResponse
+
           emit(responseCtxOut, ResponseRenderingContext(response, requestStart.method, requestStart.protocol, close),
-            pullHttpResponseIn)
-          if (close) complete(responseCtxOut)
-          // when the client closes the connection, we need to pull onc more time to get the
-          // request parser to complete
-          if (close && isEarlyResponse) pull(requestParsingIn)
+            () ⇒ {
+              if (close) {
+                complete(responseCtxOut)
+                // when the client closes the connection, we need to pull onc more time to get the
+                // request parser to complete (it will cause an incoming EntityStreamError)
+                if (requestStart.expect100Continue) pull(requestParsingIn)
```

Comment:
this changes the original condition from `close && isEarlyResponse` to `close && expect100Continue`. Is this intentional? What is the reason for the change? (The other change is that these are executed only after `emit` finishes)